### ISSUE TITLE
org.eclipse.jetty.server.ResourceCache exceptions under high load

### ISSUE THREAD

**Issue by jmcc0nn3ll** [eclipse/jetty.project#153]
migrated from Bugzilla [#441475](https://bugs.eclipse.org/bugs/show_bug.cgi?id=441475)
status ASSIGNED severity *normal* in component *server* for *9.3.x*
Reported in version *9.2.2* on platform *PC*
Assigned to: Greg Wilkins

Original attachment names and IDs:
- _[log.txt](https://bugs.eclipse.org/bugs//attachment.cgi?id=245981)_ (ID 245981)
- _[poc.tar.gz](https://bugs.eclipse.org/bugs//attachment.cgi?id=246046)_ (ID 246046)
- _[poc-0.0.1-SNAPSHOT.war](https://bugs.eclipse.org/bugs//attachment.cgi?id=246047)_ (ID 246047)
- _[logs.tar.gz](https://bugs.eclipse.org/bugs//attachment.cgi?id=246048)_ (ID 246048)

On 2014-08-11 01:17:39 -0400, Laszlo Hornyak wrote:

> When high traffic is generated against a resource served from a webjar, the ResourceCache generates some InvalidArgumentExceptions
> 
> Most frequently this is the exception:
> 
> 2014-08-10 22:39:14.502:WARN:oejs.ResourceCache:qtp1875061501-46: 
> java.lang.IllegalArgumentException: invalid resource: jar:file:/home/kocka/.m2/repository/org/webjars/angularjs/1.2.22-SNAPSHOT/angularjs-1.2.22-SNAPSHOT.jar!/META-INF/resources/webjars/angularjs/1.2.21/angular.min.js len=-1
> 	at org.eclipse.jetty.util.BufferUtil.toBuffer(BufferUtil.java:874)
> 	at org.eclipse.jetty.server.ResourceCache.getIndirectBuffer(ResourceCache.java:286)
> 	at org.eclipse.jetty.server.ResourceCache$Content.getIndirectBuffer(ResourceCache.java:441)
> 	at org.eclipse.jetty.server.ResourceCache$Content.getInputStream(ResourceCache.java:488)
> 	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:658)
> 	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:560)
> 	at org.eclipse.jetty.servlet.DefaultServlet.sendData(DefaultServlet.java:943)
> 	at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:531)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
> 	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:769)
> 	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
> 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
> 	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
> 	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
> 	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125)
> 	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
> 	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
> 	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059)
> 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
> 	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
> 	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
> 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
> 	at org.eclipse.jetty.server.Server.handle(Server.java:485)
> 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:290)
> 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)
> 	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:606)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:535)
> 	at java.lang.Thread.run(Thread.java:744)
> 
> Another, less frequent exception:
> 
> 2014-08-10 22:39:14.626:WARN:oejs.ResourceCache:qtp1875061501-73: Could not load jar:file:/home/kocka/.m2/repository/org/webjars/angularjs/1.2.22-SNAPSHOT/angularjs-1.2.22-SNAPSHOT.jar!/META-INF/resources/webjars/angularjs/1.2.21/angular.min.js true 1407696317000 java.nio.HeapByteBuffer[pos=0 lim=22 cap=22] java.nio.HeapByteBuffer[pos=0 lim=29 cap=29]
> 2014-08-10 22:39:14.831:WARN:oejs.ResourceCache:qtp1875061501-49: 
> java.io.IOException: Stream closed
> 	at java.util.zip.InflaterInputStream.ensureOpen(InflaterInputStream.java:67)
> 	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:142)
> 	at java.io.FilterInputStream.read(FilterInputStream.java:133)
> 	at java.io.FilterInputStream.read(FilterInputStream.java:133)
> 	at org.eclipse.jetty.util.BufferUtil.readFrom(BufferUtil.java:454)
> 	at org.eclipse.jetty.util.BufferUtil.toBuffer(BufferUtil.java:885)
> 	at org.eclipse.jetty.server.ResourceCache.getDirectBuffer(ResourceCache.java:303)
> 	at org.eclipse.jetty.server.ResourceCache$Content.getDirectBuffer(ResourceCache.java:463)
> 	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:640)
> 	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:560)
> 	at org.eclipse.jetty.servlet.DefaultServlet.sendData(DefaultServlet.java:943)
> 	at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:531)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
> 	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:769)
> 	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
> 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
> 	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
> 	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
> 	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125)
> 	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
> 	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
> 	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059)
> 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
> 	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
> 	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
> 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
> 	at org.eclipse.jetty.server.Server.handle(Server.java:485)
> 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:290)
> 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)
> 	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:606)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:535)
> 	at java.lang.Thread.run(Thread.java:744)

On 2014-08-13 18:42:18 -0400, Jan Bartel wrote:

> Laszlo,
> 
> What kind of load are we talking about?
> 
> Does one of these exceptions always precede the other?
> 
> Is maven likely to be updating the repository as jetty is running?
> 
> What is the configuration of your webapp?
> 
> Finally, can you provide a small test harness that reproduces the problem?

On 2014-08-14 01:23:02 -0400, Laszlo Hornyak wrote:

> Hi Jan,
> 
> 
> (In reply to Jan Bartel from comment # 1)
> > Laszlo,
> > 
> > What kind of load are we talking about?
> 
> I have generated load with apache benchmark. The exceptions only appear in the log if the requested resource is from a jar file (META-INF/resources) and I run multiple threads. For example 'ab -n 10000 -c 4 http://127.0.0.1:8080/webjars/angularjs/1.2.21/angular.min.js' in this case it generated some 10+ exceptions. 
> 
> > 
> > Does one of these exceptions always precede the other?
> 
> I do not see a clear pattern for the exceptions though, I only noticed that the first comes more often and the second comes sometimes.
> 
> > 
> > Is maven likely to be updating the repository as jetty is running?
> 
> No, the resource is downloaded and no other maven process running.
> 
> > 
> > What is the configuration of your webapp?
> > 
> > Finally, can you provide a small test harness that reproduces the problem?
> 
> I will try to isolate the problem and attach a war file. (TODO)

On 2014-08-14 01:26:17 -0400, Laszlo Hornyak wrote:

> Created attachment 245981
> log file

On 2014-08-15 12:00:56 -0400, Laszlo Hornyak wrote:

> Created attachment 246046
> maven project that produces the demonstration war file

On 2014-08-15 12:03:02 -0400, Laszlo Hornyak wrote:

> Created attachment 246047
> built war file

On 2014-08-15 12:04:09 -0400, Laszlo Hornyak wrote:

> Created attachment 246048
> jetty log files

On 2014-08-15 12:07:52 -0400, Laszlo Hornyak wrote:

> Server Software:        Jetty(9.2.2.v20140723)
> Server Hostname:        localhost
> Server Port:            8080
> 
> Document Path:          /poc/webjars/angular-ui-bootstrap/0.11.0/ui-bootstrap-tpls.min.js
> Document Length:        64353 bytes
> 
> Concurrency Level:      4
> Time taken for tests:   14.068 seconds
> Complete requests:      2000
> Failed requests:        1
>    (Connect: 0, Receive: 0, Length: 1, Exceptions: 0)
> Non-2xx responses:      1
> Total transferred:      129004131 bytes
> HTML transferred:       128644114 bytes
> Requests per second:    142.17 [#/sec] (mean)
> Time per request:       28.136 [ms] (mean)
> Time per request:       7.034 [ms] (mean, across all concurrent requests)
> Transfer rate:          8955.02 [Kbytes/sec] received
> 
> Connection Times (ms)
>               min  mean[+/-sd] median   max
> Connect:        0    0   0.3      0       9
> Processing:     5   28  27.5     21     375
> Waiting:        4   24  25.1     18     370
> Total:          5   28  27.5     21     375
> 
> Percentage of the requests served within a certain time (ms)
>   50%     21
>   66%     28
>   75%     34
>   80%     38
>   90%     51
>   95%     66
>   98%    103
>   99%    127
>  100%    375 (longest request)
> 
> 
> I suspect that something is worng with the caching, otherwise the response time should be about ten times lower. Also it may be a concurrency issue as the exceptions do not come when running ab with single thread.

On 2014-08-26 03:54:57 -0400, Laszlo Hornyak wrote:

> One possible reason is that JarResource always gets invalidated in the resource cache on each request, the next request will need to replace the resource created. This causes the low performance.

On 2014-08-26 04:20:55 -0400, Jan Bartel wrote:

> Hi Laszlo,
> 
> Yep, its because of the URLResource.getInputStream() method that nulls out the connection each time, which causes the JarFileResource to null out its fields, so getLength() will then return -1 because the _entry field is null, and thus the ResourceCache decides the JarFileResource is invalid.
> 
> I am still looking into this. I will try to come up with a bandaid solution for jetty-9.2.3, but I think the real solution is to re-architect the xxResource classes for jetty-9.3.x.
> 
> I am still testing my solution for jetty-9.2.3 - I want to make sure that it doesn't just look like its working and hasn't in fact broken somewhere else. Should have something for you in the next few days.
> 
> regards
> Jan

On 2014-08-26 07:45:49 -0400, Laszlo Hornyak wrote:

> Thank you, Jan!

On 2014-08-27 00:45:51 -0400, Jan Bartel wrote:

> Hi Lazslo,
> 
> Fixed checked in for jetty-9.2.3.
> 
> I'll leave this issue open for now until we also do the reimplementation for jetty-9.3.
> 
> cheers
> Jan

On 2015-11-13 09:44:06 -0500, Olaf Willuhn wrote:

> I think, this change (https://github.com/eclipse/jetty.project/commit/SHA: 8ce96cdd2ea5af38300b8147457e84c3eee2f48d) is the reason for a "zip file closed", i'm now getting sometimes. Once, the problem occurs, it cannot be solved using a browser refresh. I have to restart my server.
> 
> java.lang.IllegalStateException: zip file closed
>         at java.util.zip.ZipFile.ensureOpen(ZipFile.java:669)
>         at java.util.zip.ZipFile.getInputStream(ZipFile.java:351)
>         at java.util.jar.JarFile.getInputStream(JarFile.java:447)
>         at sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:162)
>         at org.eclipse.jetty.util.resource.URLResource.getInputStream(URLResource.java:241)
>         at org.eclipse.jetty.util.resource.JarResource.getInputStream(JarResource.java:122)
>         at org.eclipse.jetty.util.BufferUtil.toBuffer(BufferUtil.java:924)
>         at org.eclipse.jetty.server.ResourceCache.getDirectBuffer(ResourceCache.java:367)
>         at org.eclipse.jetty.server.ResourceCache$CachedHttpContent.getDirectBuffer(ResourceCache.java:589)
>         at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:769)
>         at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:654)
>         at org.eclipse.jetty.servlet.DefaultServlet.sendData(DefaultServlet.java:964)
>         at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:520)
>         at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
>         at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
>         at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:821)
>         at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:583)
>         at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
>         at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
>         at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
>         at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
>         at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
>         at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
>         at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
>         at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
>         at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213)
>         at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
>         at org.eclipse.jetty.server.Server.handle(Server.java:517)
>         at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:306)
>         at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
>         at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
>         at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
>         at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
>         at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
>         at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
>         at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
>         at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
>         at java.lang.Thread.run(Thread.java:745)
> 
> 
> 
> Before Jetty 9.2.3, "URLResource" dropped the "_connection" after each use.
> Now it's kept until "close()" is called explicitly.
> 
> IF this "JarURLConnection" becomes unusable (i.e. when the underlying zip file was closed for whatever reason), the "JarResource" remains in a broken state. This is because "checkConnection()" only checks, if "_connection" is not null.
> 
> My dirty fix in JarResource looks like this - it tries to "auto-repair" the connection when calling "getInputStream".
> 
> 
> private int retry = 0;
> 
> public InputStream getInputStream() throws java.io.IOException
> {     
>   try
>   {
>     checkConnection();
>     [...]
>   }
>   catch (IllegalStateException ie)
>   {
>     // zip file closed
>     if (retry++ > 3)
>       throw ie; // unable to reopen jar
> 
>     LOG.warn("jar file seems to be closed, try to re-open. message was: {}",ie.getMessage());
>     close();
>         
>     InputStream is = this.getInputStream();
>     retry = 0; // reset error count
>     return is;
>   }
> }

**Comment by billccn** [comment]
+1

Also, see my SO question with some btrace on this: https://stackoverflow.com/questions/35795256/incorrect-jarfile-sharing-in-jarurlconnection

TL;DR: The code that closes the ZipFile seems to be `JarUrlConnection` but the call seems "impossible" under normal conditions.

**Comment by gregw** [comment]
This code has been substantially reworked, so closing as old.  Please reopen if it can be reproduced on jetty 9.3.x or 9.4.x
